sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -1066,7 +1079,12 @@ public SpannerWriteResult expand(PCollection<MutationGroup> input) {
                               spec.getBatchSizeBytes(),
                               spec.getMaxNumMutations(),
                               spec.getMaxNumRows(),
-                              spec.getGroupingFactor(),
+                              // Do not group on streaming unless explicitly set.
+                              spec.getGroupingFactor()
+                                  .orElse(
+                                      input.isBounded() == IsBounded.BOUNDED
```

Comment:
I was wondering if this condition needs to be based on the input passed to this stage or based on some parameter from the user?

Reply:
it's kinda both!
If the source is unbounded (streaming) -  and the groupingFactor has not been specified by the user, then default to no grouping.

Reply:
Is there any chance that someone using SpannerIO in a streaming pipeline is relying on the default grouping factor being 1000? I'm concerned this backwards-incompatible change could break them. Would it be sufficient to just give users the _option_ to disable batching by setting the grouping factor to 1?

Reply:
They already can set groupingFactorb to 1 if they want...
Breaking backward compatibility: unlikely. 

The default of 1000 causes OOMs when using streaming, with wide windows, and high throughput... When this happens, it is not always obvious that grouping is the issue... 

With smaller windows/less throughput, it is much less likely that a group will be filled, (groups are bounded by bundles, which are bounded by windows)., So it is unlikely that anyone ever got to fill the group with 1000 batches.

Reply:
> They already can set groupingFactorb to 1 if they want...

Ha yeah sorry that was unclear. At the time I thought that groupingFactor = 1 enabled the optimization in https://github.com/apache/beam/pull/11529, so I was wondering if this was really necessary since users could just enable them by setting grouping factor manually. But I see now that grouping is separate from batching. And its disabling batching that enables your other PR.